File: Python/Program/ProgramStudi.py
# Import Class yang digunakan
from Course import Course 

class ProgramStudi:
    # Kelas ProgramStudi memiliki hubungan:
    # - Agregasi dengan Kelas SivitasAkademik
    # - Komposisi dengan Kelas Course

    # Hubungan Komposisi dengan Kelas Course digunakan karena Course atau Mata Kuliah merupakan part-of Program Studi.
    # Hal tersebut bermakna ketika Program Studi dihapus, maka Course juga akan terhapus.

    # ATRIBUT
    __nama_prodi = ""
    __kode = ""
    __courses = []

    # METHOD
    # Constructor
    def __init__(self, namaProdi = "", kode = ""):
        self.__nama_prodi = namaProdi
        self.__kode = kode

    # Setter and Getter
    def getNamaProdi(self):
        return self.__nama_prodi
    
    def setNamaProdi(self, namaProdi):
        self.__nama_prodi = namaProdi

    def getKode(self):
        return self.__kode
    
    def setKode(self, kode):
        self.__kode = kode

    # Function untuk menambahkan objek Course baru ke dalam list of Objects Course
    def addCourse(self, a:Course):
        
        listTemp = self.__courses.copy()
        listTemp.append(a)
        self.__courses = listTemp

    # Function yang mencetak isi list of Objects Courses dengan data-datanya
    def getCourses(self):
        for matkul in self.__courses:
            print ("- " + matkul.getNamaMatkul() + " (" + str(matkul.getJumlahSKS()) + " SKS)")
        print ("")

File: CPP/Program/programStudi.cpp
#include "header.hh"

// == Kelas ProgramStudi ==

// METHODS
// Constructors
ProgramStudi::ProgramStudi()
{
    this->nama_prodi = "";
    this->kode = "";
}

ProgramStudi::ProgramStudi(string nama_prodi, string kode)
{
    this->nama_prodi = nama_prodi;
    this->kode = kode;
}

// Getter and Setter
string ProgramStudi::getNamaProdi()
{
    return this->nama_prodi;
}

string ProgramStudi::getKodeProdi()
{
    return this->kode;
}

void ProgramStudi::showCourses()
{
    if (this->courses.size() == 0)
    {
        cout << "List Mata Kuliah Program Studi ini Kosong." << endl;
    }
    else
    {
        for (list<Course>::iterator it = this->courses.begin(); it != this->courses.end(); it++)
        {
            cout << "       - " << it->getNamaMatkul() << " - " << it->getJumlahSKS() << " SKS";
            cout << "\n";
        }
    }
}

list<Course> ProgramStudi::getCourses()
{
    return this->courses;
}

void ProgramStudi::setNamaProdi(string nama_prodi)
{
    this->nama_prodi = nama_prodi;
}

void ProgramStudi::setKodeProdi(string kode)
{
    this->kode = kode;
}

void ProgramStudi::addCourse(Course matkul)
{
    this->courses.push_back(matkul);
    // this->showCourses();
}

// Destructor
ProgramStudi::~ProgramStudi()
{
}
